Brokerages/WebSocketWrapper.cs

```diff
@@ -63,17 +55,110 @@ public void Initialize(string url)
+                return;
+            }
+
+            try
+            {
+                _wrapped.ConnectAsync(new Uri(_url), _cancellationTokenSource.Token).SynchronouslyAwaitTask();
+
+                if (IsOpen)
+                {
+                    OnOpen();
+
+                    new TaskFactory().StartNew(
+                        async () =>
+                        {
+                            try
+                            {
+                                await StartReceiving();
+                            }
+                            catch (Exception exception)
+                            {
+                                OnError(new WebSocketError(exception.Message, exception));
+                            }
+
+                            if (_wrapped.State == WebSocketState.Aborted)
+                            {
+                                // need to recreate ClientWebSocket instance, cannot be reused
+
+                                Log.Trace($"WebSocketWrapper.Connect(): Connection aborted (IsOpen:{IsOpen}, State:{_wrapped.State}): creating new ClientWebSocket instance", true);
+
+                                await new TaskFactory().StartNew(ResetAndConnect, _cancellationTokenSource.Token);
+                            }
+
+                        }, _cancellationTokenSource.Token);
+                }
+            }
+            catch (Exception exception)
             {
-                _wrapped.Connect();
+                OnError(new WebSocketError(exception.Message, exception));
+            }
+        }
+
+        private void ResetAndConnect()
+        {
+            _wrapped = new ClientWebSocket();
+            Connect();
+        }
+
+        private async Task StartReceiving()
+        {
+            while (_wrapped.State == WebSocketState.Open)
+            {
+                using (var stream = new MemoryStream(1024))
+                {
+                    WebSocketReceiveResult webSocketReceiveResult;
+                    do
+                    {
+                        webSocketReceiveResult = await _wrapped.ReceiveAsync(_receiveBuffer, _cancellationTokenSource.Token);
+                        await stream.WriteAsync(_receiveBuffer.Array, _receiveBuffer.Offset, webSocketReceiveResult.Count, _cancellationTokenSource.Token);
+                    } while (!webSocketReceiveResult.EndOfMessage);
+
+                    if (webSocketReceiveResult.MessageType == WebSocketMessageType.Text)
+                    {
+                        stream.Seek(0, SeekOrigin.Begin);
+                        using (var reader = new StreamReader(stream, Encoding.UTF8))
+                        {
+                            OnMessage(new WebSocketMessage(reader.ReadToEnd()));
+                        }
+                    }
+                    else if (webSocketReceiveResult.MessageType == WebSocketMessageType.Close)
+                    {
+                        if (webSocketReceiveResult.CloseStatus != null)
```

Comment:
Maybe we would want to send the `OnClose` event even if we don't have access to the `CloseStatus`?